week-02-ajax-and-js-objects/day-04/module-01/readme.md

```diff
@@ -0,0 +1,179 @@
+
+*Spoiler* `this` went from being the `window` or global state to being an object's internals. `this` changes around because of the `new` keyword and how `this` is `bound`. When we use the `new` keyword we follow a path similar to <a href="http://eddmann.com/posts/how-the-new-keyword-works-in-javascript/" target="_blank">this</a> and <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new" target="_blank">this</a>.
+
+* We create an object, a more formal way of doing `{}` and give it a default set of information.
+  * `var obj = Object.create(Carpet.prototype);`
+* We `apply` our new `obj` to the original `function` which changes `this` to point at `obj` instead of `window`.
+  * `var ret = Carpet.apply(obj);`
+
+Aha! Now we see the `prototype`, it is behind the scenes of `new`.
+
+**Warning** side note on `.apply`. You'll often see `.apply` and `.call` in JavaScript. They are functions which change what `this` points to.
+
+```js
+function fly() {
+  console.log(this);
+  if (this.magic) {
+    console.info("We're flying!");
+  } else {
+    console.warn("Falling.");
+  }
+}
+
+fly(); // Falling.
+
+var magicCarpet = {"magic": true};
+fly.call(magicCarpet) // We're flying!
+```
+
+## Using prototypes to make a library
+
+Although prototypes are designed as part of the OO nature of JavaScript, they aren't as often used in the same way. Recently, this is becoming a common trend with different JavaScript frameworks including their own Data Abstraction Layer.
+
+So why are prototypes important?
+
+Because we use JavaScript frameworks and libraries often. Libraries like jQuery, Bootstrap or countless others. Often the best way to quickly implement something for your site is via an existing open source library. Understanding how these libraries work allow us to extend and fix their functionality.
+
+Let's make a small library together.
+
+## Chaos Library v0.2.a-1
+
+We're going to make a chaotic library to highlight using prototypes in extending the library.
+
+```js
+var Chaos = function Chaos(level) {
+  this.level = level;
+}
+
+Chaos.prototype.havoc = function () {
+  if (this.level > 2) {
+    console.warn("We're causing havoc!");
+  } else {
+    console.info("No havoc, let's take a break.");
+  }
+}
+```
+
+Now I can try out my little library.
+
+```js
+var myChaos = new Chaos(1);
+myChaos.havoc();
+```
+
+This is great but I actually want all levels of chaos to trigger havoc. In order to do this, I *only* need to change its prototype.
+
+```js
+Chaos.prototype.havoc = function () {
+  console.warn("WE ALWAYS CAUSE HAVOC!");
+}
+```
+
+What will happen when I run `myChaos.havoc()` without creating it again?
+
+```js
+myChaos.havoc()
+```
+
+It causes havoc now. This means that anytime I edit the prototype for a type of object it will change all the instances of that object immediately. It will also change the behaviour of new instances.
```

Comment:
This is a great explanation - very clear. I would also mention something about how we're saving memory by storing functionality on the prototype as opposed to the constructor.